src/main/kotlin/io/github/mojira/arisa/ArisaMain.kt

```diff
@@ -1,165 +1,68 @@
-
-        if (executionResults.successful) {
-            rerunTickets = emptySet()
-            failedTickets.addAll(executionResults.failedTickets)
-            val failed = failedTickets.joinToString("") { ",$it" } // even first entry should start with a comma
-
-            if (config[Arisa.Debug.updateLastRun]) {
-                lastRunFile.writeText("${currentRunTime.toEpochMilli()}$failed")
-            }
-            lastRunTime = currentRunTime
-            TimeUnit.SECONDS.sleep(checkIntervalSeconds)
-        } else {
-            if (Duration.between(lastRelog, Instant.now()).toMinutes() >= 1) {
-                // If last relog was more than a minute before and execution failed with an exception, relog
-                log.info("Trying to relog")
-                try {
-                    jiraClient =
-                        connectToJira(
-                            config[Arisa.Credentials.username],
-                            config[Arisa.Credentials.password],
-                            config[Arisa.Issues.url]
-                        )
-                    log.info("Relog was successful")
-                    lastRelog = Instant.now()
-                    executor = Executor(
-                        config, moduleRegistries, getSearchIssues(jiraClient, config)
-                    )
-                    TimeUnit.SECONDS.sleep(checkIntervalSeconds)
-                } catch (exception: Exception) {
-                    log.error("Relog failed", exception)
-                    // Wait before performing any further action
-                    TimeUnit.SECONDS.sleep(max(RELOG_INTERVAL_IN_SECONDS, checkIntervalSeconds))
-                }
-            } else {
-                // Not enough time for relog passed, just try waiting a bit
-                log.info("Not enough time for relog has passed")
-                TimeUnit.SECONDS.sleep(max(WAIT_TIME_AFTER_CONNECTION_ERROR_IN_SECONDS, checkIntervalSeconds))
-            }
-        }
+        HelperMessageUpdateService.checkForUpdate()
 
-        // Update helper messages if necessary
-        if (currentTime.epochSecond - helperMessagesLastFetch.epochSecond >= helperMessagesInterval) {
-            HelperMessageService.updateHelperMessages(helperMessagesFile)
-            executor = Executor(
-                config, moduleRegistries, getSearchIssues(jiraClient, config)
-            )
-            helperMessagesLastFetch = currentTime
-        }
+        val secondsToSleep = runExecutionCycle(executor)
+        TimeUnit.SECONDS.sleep(secondsToSleep)
     }
 }
 
-private fun getSearchIssues(
-    jiraClient: JiraClient,
-    config: Config
-): (String, Int, () -> Unit) -> List<Issue> {
-    return ::searchIssues
-        .partially1(jiraClient)
-        .partially1(config)
-}
+/**
+ * @return amount of seconds to sleep after this execution cycle
+ */
+private fun runExecutionCycle(executor: Executor): Long {
+    val timeframe = ExecutionTimeframe.getTimeframeFromLastRun()
```

Comment:
maybe this whole loop can be a ExecutionService and Arisa just initializes it and calls a run on it?

Reply:
The idea is that the main is as small as possible

Reply:
I tried extracting this into another file. The problem is, this function right here relies on quite a lot of stuff in `ArisaMain`, like `config`, `connectionService`, and `executor`. I didn't think it was worth it, and having all the "glue" here in one place is pretty good imo.